sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,687 @@
+        might change in the future for a better heuristic.
+
+    bisect_strategy : {"biggest_sse", "child_biggest_sse", "largest_cluster"},
+        default="biggest_sse"
+        Defines how should bisection by performed.
+        - "biggest_sse" means that Bisect K-Means will always check
+        all calculated cluster for cluster with biggest SSE
+        (Sum of squared errors) and bisect it. That way calculated clusters
+        will be more balanced
+        - "child_biggest_sse" means that Bisect K-Means will always check
+        SSE of only clusters obtained from previous iteration for bisection.
+        Calculated clusters will be less balanced - consecutive clusters
+        will be usually smaller than previous
+        - "largest_cluster" - Bisect K-Means will always split cluster with
+        largest amount of points assigned to it from all clusters
+        previously calculated. That should work faster than picking by SSE
+        ('biggest_sse') and may produce similar results in most cases
+
+
+    Attributes
+    ----------
+    cluster_centers_ : ndarray of shape (n_clusters, n_features)
+        Coordinates of cluster centers. If the algorithm stops before fully
+        converging (see ``tol`` and ``max_iter``), these will not be
+        consistent with ``labels_``.
+
+    labels_ : ndarray of shape (n_samples,)
+        Labels of each point
+
+    inertia_ : float
+        Sum of squared distances of samples to their closest cluster center,
+        weighted by the sample weights if provided.
+
+    n_iter_ : int
+        Number of iterations run.
+
+    Notes
+    -----
+    That algorithm will not work if n_cluster is smaller than 2.
+
+    Also it might be inefficient when n_cluster is equal to 2
+
+    Examples
+    --------
+    >>> from sklearn.cluster import BisectKMeans
+    >>> import numpy as np
+    >>> X = np.array([[1, 2], [1, 4], [1, 0],
+    ...               [10, 2], [10, 4], [10, 0],
+    ...               [10, 6], [10, 8], [10, 10]])
+    >>> bisect_means = BisectKMeans(n_clusters=3, random_state=0).fit(X)
+    >>> bisect_means.labels_
+    array([0, 0, 0, 2, 2, 2, 1, 1, 1], dtype=int32)
+    >>> bisect_means.predict([[0, 0], [12, 3]])
+    array([0, 2], dtype=int32)
+    >>> bisect_means.cluster_centers_
+    array([[ 1.,  2.],
+           [10.,  8.],
+           [10.,  2.]])
+    """
+    def __init__(self,  n_clusters=8, init='k-means++', n_init=10,
+                 random_state=None, max_iter=30, verbose=0,
+                 tol=1e-4, copy_x=True, algorithm='auto',
+                 bisect_strategy='biggest_sse'):
+
+        super().__init__(
+            n_clusters=n_clusters, init=init, max_iter=max_iter,
+            verbose=verbose, random_state=random_state, tol=tol,
+            n_init=n_init, copy_x=copy_x, algorithm=algorithm)
+
+        self.bisect_strategy = bisect_strategy
+
+    def _compute_bisect_errors(self, X, centers, labels):
```

Comment:
It would be nice to handle sample weights here.

Reply:
Will be added along with`_inerita_sparse` and `_inertia_dense`